### PR TITLE
[HM-02] By Mikhail Sauchuk

### PR DESCRIPTION
**Pull request by misha-sauchuk**
@vladhilko @B0RlS @alexshgov

This PR includes four tasks from HM-02 and two _.log_ files for **task_02** and **task_03**

### CODE REVIEW

File: HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_02.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+# Method `task_2` should return:
+# an array of formatted strings containing information about post requests;
+# for the example above, the result should be:
+# lines that do not match the format:
+# 10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /test/2/messages HTTP/1.1"
+# should be ignored
```
Comment (vladhilko) [inline HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_02.rb]: Looks like you forgot to check some fields 😅 
```suggestion
# 10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /test/2/messages HTTP/1.1" 200 48 0.0498"
# should be ignored
```

File: HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_02.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+def main
+  puts 'Please enter the full path to log file you would like to chekc or ' \
+    'press Enter to test the code'
+  user_input = gets.chomp
+  file_name = check_input(user_input)
+  search_pattern = "#{IP_PATTERN} - - \\[#{DATE_PATTERN}\\] #{METHOD_PATTERN}.*"
+  post_messages = find_post_messages(file_name, search_pattern)
+  get_info(post_messages)
+end
```
Comment (vladhilko) [inline HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_02.rb]: All your methods should be under the `main` method

File: HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_01.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+# - full text of the first line with an error
+# (the presence of the substring `error` in any case)
+# - If the input is text that does not contain the word `error`,
+# return an empty string
+
+LOG_FILE = 'bash.log'
+
+def find_error(file_name)
+  # method rerturns first line with error from the input log
+  File.open(file_name, 'r') do |f|
```
Comment (vladhilko) [inline HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_01.rb]: You can use one line syntax 
```suggestion
  File.readlines(file_name)
```

File: HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_02.rb (hunk 1 above)
Comment (misha-sauchuk) [reply HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_02.rb]: I have added this line to the description. Also script checks the existence of this part in the line

File: HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_02.rb (hunk 2 above)
Comment (misha-sauchuk) [reply HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_02.rb]: I have moved `main` method to the top

File: HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_01.rb (hunk 3 above)
Comment (misha-sauchuk) [reply HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_01.rb]: I have refactored three tasks with your suggestions, thanks

File: HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_01.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    'press Enter to test the code'
+  user_input = gets.chomp
+  log_to_check = check_input(user_input)
+  find_error(log_to_check)
+end
+
+def find_error(file_name)
+  # method rerturns first line with error from the input log
+  lines = File.readlines(file_name)
+  error_lines = lines.find_all { |line| line.downcase.include? 'error' }
```
Comment (vladhilko) [inline HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_01.rb]: We can use `find` to return only one record
```suggestion
  error_lines = lines.find { |line| line.downcase.include? 'error' }
```

Comment (misha-sauchuk) [reply HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_01.rb]: Replaced find_all with find method

File: HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_04.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# An arbitrary string is given, the method `task_4`
+# must return the count of letters and digits in that string.
+
+def count_letters(str)
```
Comment (vladhilko) [inline HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_04.rb]: Sounds more clear for me
```suggestion
def letters_count(str)
```

File: HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_04.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  str.scan(/[a-zA-Z]/).size
+end
+
+def count_digits(str)
+  str.scan(/[0-9]/).size
+end
+
+def main(str)
+  letters = count_letters(str)
+  digits = count_digits(str)
```
Comment (vladhilko) [inline HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_04.rb]: these variables aren't necessary, we can use these methods inside our `puts` explicitly, but it's minor

File: HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_04.rb (hunk 5 above)
Comment (misha-sauchuk) [reply HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_04.rb]: Got it, and have renamed

File: HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_04.rb (hunk 6 above)
Comment (misha-sauchuk) [reply HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_04.rb]: ok, thanks

File: HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_01.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+# (the presence of the substring `error` in any case)
+# - If the input is text that does not contain the word `error`,
+# return an empty string
+
+LOG_FILE = 'bash.log'
+
+def main
+  puts 'Please enter the full path to log file you would like to chekc or ' \
+    'press Enter to test the code'
+  user_input = gets.chomp
```
Comment (vladhilko) [inline HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_01.rb]: Normally, we don't use `gets.chomp` at all in real production code, you don't need to worry about this input. Would be better to think about creating a clear, readable and easy solution

Comment (misha-sauchuk) [reply HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_01.rb]: Should I do any action on this comment?

Comment (vladhilko) [reply HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_01.rb]: Whatever you want, just don't add it in the future

Comment (misha-sauchuk) [reply HW02/Mikhail_Sauchuk/mikhail_sauchuk_hw_02_t_01.rb]: ok, got it!

**Pull request merged**